src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
   | EvalErrorType TypeError
   | EvalErrorCase SrcLoc Value
   | EvalErrorUser Syntax
-  deriving (Show)
+  | EvalErrorIdent Value
+  deriving (Show, Typeable)
 makePrisms ''EvalError
+instance Exception EvalError
+
+-- | The Kontinuation type. The naming convention InFoo means that the subject
+-- of evaluation in the CEK machine is Foo. For example, when the continuation
+-- is 'InArg' the subject of evaluation is the argument of the function and the
+-- continuation holds the un-evaluated function symbol in its first field.
+data Kont where
+
+  Halt :: Kont
+  -- ^ Marks the evaluator finishing
+
+  -- functions
+  InArg :: !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  -- ^ The argument is being evaluated, so hold onto the function symbol.
+  InFun :: !Value -> !VEnv -> !Kont -> Kont
+  -- ^ The function is being evaluated, so hold onto the evaluated argument.
```

Comment:
Hmm, you evaluate the argument first? I of course know that Klister is a strict language, which means that the arguments are evaluated before the call is made, but what about the case in which the function is an expression, e.g. `((id (+)) (* 2 2) (* 3 3))`? Isn't it standard to evaluate `(id (+))` to `(+)` before evaluating `(* 2 2)` to `4` and `(* 3 3)` to `9`?